lib/rouge/lexer.rb

```diff
@@ -423,15 +423,22 @@ def reset!
     #
     # @note The use of `opts` has been deprecated. A warning is issued if run
     #   with `$VERBOSE` set to true.
-    def lex(string, opts={}, &b)
-      unless opts.nil?
-        warn 'The use of opts with Lexer.lex is deprecated' if $VERBOSE
+    def lex(string, opts=nil, &b)
+      if opts
+        if opts.empty?
+          warn 'the use of options with Lexer#lex is deprecated' if $VERBOSE
+        elsif opts.size == 1 && opts[:continue]
+          warn '`lex :continue => true` is deprecated, please use #continue_lex instead'
+          return continue_lex(string, &b)
+        else
+          raise ArgumentError.new("invalid options for Lexer#lex: #{opts.inspect}")
```

Comment:
I don't think throwing an exception, and forcing the calling application to crash, is the right move here. (And it will crash past releases of Asciidoctor PDF). I think to start, this should be the same as the first condition, which is the warn in verbose mode and otherwise ignore the options.

In other words:

```rb
if opts
  if opts.size == 1 && opts[:continue]
    warn '`lex :continue => true` is deprecated, please use #continue_lex instead'
    return continue_lex(string, &b)
  else
    warn 'the use of options with Lexer#lex is deprecated' if $VERBOSE
  end
end
```

Yes, it was a mistake for Asciidoctor & Asciidoctor PDF to pass options to this method. But the reality is that the code is out there already and we're going to break builds by throwing an exception (at least until Asciidoctor, Asciidoctor PDF, AsciidoctorJ, and AsciidoctorJ PDF all have a chance to release an update).

Reply:
Do current releases of Asciidoctor not lock down the rouge version in a Gemfile.lock or similar?

Reply:
Rouge is not a hard dependency of Asciidoctor (since not all users use Rouge for syntax highlighting). Rouge must be installed by the user, which then activates the integration. So no, Asciidoctor doesn't specify the version. We don't control the user's Gemfile. (I don't know of a way to control the version of an optional dependency with RubyGems, which is why we can't specify it in the gemfile as a runtime dependency).

Reply:
I don't understand why what I'm proposing is a problem. It communicates exactly the same information without forcing a crash. It's a very fair trade-off.

For the record, what I'm proposing is almost exactly the change I proposed for 3.4.1. Had that been accepted as is, we wouldn't have had the problem we had with 3.4.1. See https://github.com/rouge-ruby/rouge/pull/1176/commits/454b185986c5956e7fd37d7f708c8ee85edbcc49. But I do like the idea of adding the warning in verbose mode, which is the compromise this code proposes:

```rb
if opts
  if opts.size == 1 && opts[:continue]
    warn '`lex :continue => true` is deprecated, please use #continue_lex instead'
    return continue_lex(string, &b)
  elsif $VERBOSE
    warn 'the use of options with Lexer#lex is deprecated'
  end
end
```

The user gets the warning, but the calling application doesn't crash.